Implement Kopia

perfectra1n opened the pull request:
**Describe what this PR does**
This PR implements Kopia as a mover.

**Is there anything that requires special attention?**
To give this a spin yourself to see if/how it works, you can review the top of the fork's [README](https://github.com/perfectra1n/volsync). You can also see the deployed Kopia documentation [here](https://perfectra1n.github.io/volsync/usage/kopia/index.html). I have a handful of commits in this branch so that the fork works as expected, but I've reverted them so that this branch can be used for the explicit purpose of merging.

**Related issues:**
Closes #320 

# Kopia
Unlike the other mover tools in VolSync, Kopia operates as a **content-addressable backup system** with fundamentally different approaches to data storage, deduplication, and repository management.

**How Kopia works:**
```
All Backups: [Content Hash ABC123] [Content Hash DEF456] [Content Hash GHI789]
                     ↑                    ↑                    ↑
Backup 1 Index: → ABC123 (File A) → DEF456 (File B) → GHI789 (File C)
Backup 2 Index: → ABC123 (File A) → DEF456 (File B) → JKL012 (File C')  # Only new content stored
Backup 3 Index: → ABC123 (File A) → MNO345 (File B') → JKL012 (File C') # Even more deduplication
```

Kopia automatically deduplicates identical content across all backups, while Restic/Rsync store incremental changes.

**How Concurrent Access Works**

**Kopia's approach:**
```
Client A ──┐
Client B ──┼─► Kopia Repository ◄─── Safe concurrent writes
Client C ──┘    (Content-addressed storage prevents conflicts)
```

Kopia's content-addressable design means multiple clients can write simultaneously because identical content gets the same hash - no lock conflicts.

**Kopia's multi-tenant design:**
```go
// Each backup gets unique identity in shared repository
username: "team-frontend"     // Tenant identifier
hostname: "prod-cluster-db"   // Application identifier

// Repository structure:
repository/
├─ snapshots/
│  ├─ team-frontend@prod-cluster-db/  # Isolated namespace
│  ├─ team-backend@prod-cluster-app/   # Different tenant
│  └─ team-data@prod-cluster-db/       # Different app, same tenant
├─ content/ (shared - deduplication across tenants!)
└─ policies/ (per-tenant policies)
```

Benefits include shared storage with isolation, cross-tenant deduplication saves space, and per-tenant policies and access control.


Kopia uses **pluggable storage drivers** that abstract the underlying storage while providing consistent repository semantics:

```
Application Layer:     [Snapshot Management] [Deduplication] [Compression]
                                      ↕
Storage Abstraction:         [Kopia Storage Interface]
                                      ↕
Storage Drivers:       [S3] [GCS] [Azure] [B2] [SFTP] [WebDAV] [FS] [Rclone]
                        ↕     ↕      ↕       ↕      ↕       ↕      ↕      ↕
Physical Storage:   [AWS] [Google] [Azure] [B2] [SSH] [WebDAV] [NFS] [100+ via Rclone]
```

**Unified Repository Format**
```yaml
# Same repository structure works across ALL backends:
KOPIA_REPOSITORY: s3://bucket/backups        # S3
KOPIA_REPOSITORY: gcs://bucket/backups       # Google Cloud  
KOPIA_REPOSITORY: azure://container/backups  # Azure
KOPIA_REPOSITORY: sftp://server/path         # SFTP
# ... exact same backup format and features regardless of backend!
```

**Native Cloud Integration**
Unlike Restic which uses generic S3 API, Kopia has **native drivers** for S3, GCS, and Azure (as examples)

**S3 Intelligent Path Parsing:**
```yaml
# Kopia automatically handles complex nested paths
KOPIA_REPOSITORY: s3://my-bucket/team-a/prod/databases/postgres
# ↓ Kopia parses this as:
# Bucket: my-bucket
# Prefix: team-a/prod/databases/postgres
# Features: Lifecycle policies, versioning, encryption all work
```

**Multi-Backend Credential Flexibility:**
```yaml
# Same secret can contain credentials for multiple backends
apiVersion: v1
kind: Secret
stringData:
  # Primary backend
  KOPIA_REPOSITORY: s3://primary-bucket/backups
  AWS_ACCESS_KEY_ID: primary-key
  
  # Fallback backend (Kopia can migrate between backends)
  KOPIA_FALLBACK_REPOSITORY: gcs://fallback-bucket/backups  
  GOOGLE_APPLICATION_CREDENTIALS: |
    {"type": "service_account", ...}
```

**Configuration Options**

The `ReplicationSource` spec includes Kopia-specific options:

```yaml
apiVersion: volsync.backube/v1alpha1
kind: ReplicationSource
metadata:
  name: database-backup
spec:
  sourcePVC: my-database
  kopia:
    repository: kopia-config
    # Performance tuning
    compression: zstd              # zstd, gzip, s2, none
    parallelism: 4                 # Parallel upload streams
    
    # Retention policy
    retain:
      hourly: 24                   # Keep 24 hourly backups
      daily: 7                     # Keep 7 daily backups  
      weekly: 4                    # Keep 4 weekly backups
      monthly: 12                  # Keep 12 monthly backups
      yearly: 5                    # Keep 5 yearly backups
    
    # Cache configuration (flexible EmptyDir fallback)
    cacheCapacity: 8Gi             # Cache size
    cacheStorageClassName: fast-ssd # Optional PVC cache
    cacheAccessModes: [ReadWriteOnce]
    
    # Multi-tenancy
    username: team-frontend        # Custom username
    hostname: prod-cluster-db      # Custom hostname
    
    # Application consistency  
    actions:
      beforeSnapshot: "pg_start_backup('volsync')"
      afterSnapshot: "pg_stop_backup()"
    
    # Path preservation
    sourcePathOverride: /var/lib/postgresql/data
    
    # Maintenance
    maintenanceIntervalDays: 7     # Run maintenance weekly
```

**Smart Cache Strategy Selection:**

Instead of requiring users to understand cache implications, the mover **automatically chooses** the best strategy:

```yaml
# Strategy 1: No cache config → EmptyDir fallback (good for testing)
spec:
  kopia: {}
# Result: EmptyDir with 8Gi limit, reasonable performance

# Strategy 2: Size-only config → EmptyDir with limit (good for constrained resources)  
spec:
  kopia:
    cacheCapacity: 4Gi
# Result: EmptyDir with 4Gi limit, predictable resource usage

# Strategy 3: Full config → Dedicated PVC (best for production)
spec:
  kopia:
    cacheCapacity: 20Gi
    cacheStorageClassName: fast-nvme
# Result: Persistent PVC, maximum performance
```

The mover implements **lifecycle-based metrics collection** throughout the backup process:

```go
func (m *Mover) Synchronize(ctx context.Context) (mover.Result, error) {
    // 1. Record operation start and set connectivity
    operationStart := time.Now()
    m.metrics.RepositoryConnectivity.With(labels).Set(1)
    
    // 2. Track cache configuration decisions
    m.recordCacheMetrics()  // Records cache type (PVC vs EmptyDir) and size
    
    // 3. Monitor job execution and retries
    if job.Status.Failed > 0 {
        m.recordJobRetry(operation, "job_pod_failure")
    }
    
    // 4. Record final outcome with duration
    if result.Completed {
        duration := time.Since(operationStart)
        m.recordOperationSuccess(operation, duration)
        // Also record maintenance operations when applicable
        if m.shouldRunMaintenance() {
            m.recordMaintenanceOperation()
        }
    } else {
        m.recordOperationFailure(operation, "job_execution_failed")
        m.metrics.RepositoryConnectivity.With(labels).Set(0)
    }
}
```

**How metrics are structured with labels:**

```go
func (m *Mover) getMetricLabels(operation string) prometheus.Labels {
    return prometheus.Labels{
        "obj_name":      m.owner.GetName(),        // ReplicationSource name
        "obj_namespace": m.owner.GetNamespace(),   // Namespace
        "role":          "source",                  // source vs destination
        "operation":     operation,                 // backup, restore, maintenance
        "repository":    m.repositoryName,         // Repository secret name
    }
}
```

**Available metrics categories:**
- **Performance**: `backup_duration_seconds`, `compression_ratio`, `data_transfer_rate`
- **Reliability**: `backup_success_total`, `backup_failure_total`, `job_retries_total`
- **Configuration**: `cache_type` (pvc/emptydir), `cache_size_bytes`, `policy_compliance`
- **Repository Health**: `repository_connectivity`, `maintenance_operations_total`

**Policy Configuration**

Support for **structured repository configuration** and **policy files**:

```yaml
kopia:
  # Option 1: Structured JSON config (validated)
  policyConfig:
    repositoryConfig: |
      {
        "splitters": {"default": "DYNAMIC-4M-BUZHASH"},
        "compression": {"compressor": "zstd", "level": 3},
        "encryption": {"algorithm": "CHACHA20-POLY1305"},
        "caching": {"maxCacheSizeBytes": 1073741824}
      }
  
  # Option 2: File-based policies from ConfigMap/Secret  
  policyConfig:
    configMapName: kopia-policies
    globalPolicyFilename: global-policy.json
    repositoryConfigFilename: repo.config
```

**Core Implementation:**
- `internal/controller/mover/kopia/mover.go` - Main mover implementation
- `internal/controller/mover/kopia/metrics.go` - Prometheus metrics
- `internal/controller/mover/kopia/builder.go` - Builder pattern for mover creation
- `mover-kopia/entry.sh` - Container entry point with debug support
- `mover-kopia/Dockerfile` - Kopia container with all backends

**CRD Extensions:**
- `api/v1alpha1/replicationsource_types.go` - ReplicationSourceKopiaSpec
- `api/v1alpha1/replicationdestination_types.go` - ReplicationDestinationKopiaSpec  
- `api/v1alpha1/common_types.go` - KopiaPolicySpec

**Documentation:**
- `docs/usage/kopia/index.rst` - User guide
- `docs/usage/kopia/database_example.rst` - Database backup examples
- Example PrometheusRule for monitoring

**Flexible Job Configuration:**
```go
func (m *Mover) ensureJob(ctx context.Context, ...) (*batchv1.Job, error) {
    // Supports EmptyDir OR PVC cache seamlessly
    // Mounts credentials for multiple backends  
    // Configures custom CA certificates
    // Sets up policy configuration files
    // Handles privileged vs unprivileged modes
}
```

**Multi-Backend Environment Variables:**
```go
func (m *Mover) buildEnvironmentVariables(repo *corev1.Secret) []corev1.EnvVar {
    envVars := m.buildBasicEnvironmentVariables()
    envVars = append(envVars, m.buildRepositoryEnvironmentVariables(repo)...)
    envVars = append(envVars, m.buildAWSEnvironmentVariables(repo)...)
    envVars = append(envVars, m.buildAzureEnvironmentVariables(repo)...)
    envVars = append(envVars, m.buildGoogleEnvironmentVariables(repo)...)
    envVars = append(envVars, m.buildB2EnvironmentVariables(repo)...)
    envVars = append(envVars, m.buildWebDAVEnvironmentVariables(repo)...)
    envVars = append(envVars, m.buildSFTPEnvironmentVariables(repo)...)
    envVars = append(envVars, m.buildRcloneEnvironmentVariables(repo)...)
    // ... additional configuration
}
```

**Database Backup:**
```yaml
apiVersion: volsync.backube/v1alpha1
kind: ReplicationSource
metadata:
  name: postgres-backup
spec:
  sourcePVC: postgres-data
  trigger:
    schedule: "0 2 * * *"  # Daily at 2 AM
  kopia:
    repository: postgres-backup-config
    retain:
      daily: 30
      weekly: 12
    actions:
      beforeSnapshot: "pg_start_backup('volsync')"
      afterSnapshot: "pg_stop_backup()"
```

**Multi-Cloud with Policies:**
```yaml
apiVersion: volsync.backube/v1alpha1
kind: ReplicationSource
metadata:
  name: app-backup-prod
spec:
  sourcePVC: app-data
  trigger:
    schedule: "0 */6 * * *"  # Every 6 hours
  kopia:
    repository: s3-backup-config
    compression: zstd
    parallelism: 8
    cacheCapacity: 32Gi
    cacheStorageClassName: fast-nvme
    username: production-app
    hostname: prod-cluster-east
    retain:
      hourly: 48
      daily: 30
      weekly: 26
      monthly: 24
      yearly: 7
    policyConfig:
      repositoryConfig: |
        {
          "compression": {"compressor": "zstd", "level": 9},
          "encryption": {"algorithm": "CHACHA20-POLY1305"},
          "splitters": {"default": "DYNAMIC-4M-BUZHASH"}
        }
```

**Restore Example:**
```yaml
apiVersion: volsync.backube/v1alpha1
kind: ReplicationDestination
metadata:
  name: restore-from-backup
spec:
  kopia:
    repository: s3-backup-config
    restoreAsOf: "2024-01-15T10:30:00Z"  # Point-in-time restore
    shallow: 1                           # Only restore recent data
  volumeSnapshotClassName: csi-snapclass
  capacity: 100Gi
```

dependabot[bot] committed:
Bump golang.org/x/net from 0.26.0 to 0.38.0 in /custom-scorecard-tests

Bumps [golang.org/x/net](https://github.com/golang/net) from 0.26.0 to 0.38.0.
- [Commits](https://github.com/golang/net/compare/v0.26.0...v0.38.0)

---
updated-dependencies:
- dependency-name: golang.org/x/net
  dependency-version: 0.38.0
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update codecov/codecov-action action to v5.4.3

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update module github.com/go-logr/logr to v1.4.3

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update syncthing packages to v1.29.7

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update k8s.io/utils digest to 4c0f3b2

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

dependabot[bot] committed:
Bump requests from 2.32.0 to 2.32.4 in /test-e2e

Bumps [requests](https://github.com/psf/requests) from 2.32.0 to 2.32.4.
- [Release notes](https://github.com/psf/requests/releases)
- [Changelog](https://github.com/psf/requests/blob/main/HISTORY.md)
- [Commits](https://github.com/psf/requests/compare/v2.32.0...v2.32.4)

---
updated-dependencies:
- dependency-name: requests
  dependency-version: 2.32.4
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

tesshuflower committed:
update changelog for v0.13.0-rc.3

Signed-off-by: Tesshu Flower <tflower@redhat.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

tesshuflower committed:
use go 1.23 for cust scorecard tests

Signed-off-by: Tesshu Flower <tflower@redhat.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

tesshuflower committed:
helm chart version to v0.13.0 in prep for release

Signed-off-by: Tesshu Flower <tflower@redhat.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update pre-commit hook renovatebot/pre-commit-hooks to v40.57.1

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

tesshuflower committed:
stop ffwding main to release-0.13

Signed-off-by: Tesshu Flower <tflower@redhat.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

dependabot[bot] committed:
Bump urllib3 from 2.4.0 to 2.5.0 in /test-e2e

Bumps [urllib3](https://github.com/urllib3/urllib3) from 2.4.0 to 2.5.0.
- [Release notes](https://github.com/urllib3/urllib3/releases)
- [Changelog](https://github.com/urllib3/urllib3/blob/main/CHANGES.rst)
- [Commits](https://github.com/urllib3/urllib3/compare/2.4.0...2.5.0)

---
updated-dependencies:
- dependency-name: urllib3
  dependency-version: 2.5.0
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

dependabot[bot] committed:
Bump github.com/go-viper/mapstructure/v2 from 2.2.1 to 2.3.0

Bumps [github.com/go-viper/mapstructure/v2](https://github.com/go-viper/mapstructure) from 2.2.1 to 2.3.0.
- [Release notes](https://github.com/go-viper/mapstructure/releases)
- [Changelog](https://github.com/go-viper/mapstructure/blob/main/CHANGELOG.md)
- [Commits](https://github.com/go-viper/mapstructure/compare/v2.2.1...v2.3.0)

---
updated-dependencies:
- dependency-name: github.com/go-viper/mapstructure/v2
  dependency-version: 2.3.0
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update pre-commit hook renovatebot/pre-commit-hooks to v41

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update dependency golangci/golangci-lint to v2.2.1

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update dependency kubernetes-sigs/kustomize to v5.7.0

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update actions/download-artifact action to v4.3.0

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update pre-commit hook renovatebot/pre-commit-hooks to v41.23.4

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update actions/setup-go action to v5.5.0

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update syncthing packages to v1.30.0

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update dependency pipenv to v2025

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

tesshuflower committed:
allow utils as pkg name for linter

Signed-off-by: Tesshu Flower <tflower@redhat.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update pre-commit hook renovatebot/pre-commit-hooks to v41.23.5

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

tesshuflower committed:
versioning for next release v0.14, changelog etc

Signed-off-by: Tesshu Flower <tflower@redhat.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

tesshuflower committed:
update ffwd from main->release0.14

Signed-off-by: Tesshu Flower <tflower@redhat.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

dependabot[bot] committed:
Bump golang.org/x/oauth2 in /custom-scorecard-tests

Bumps [golang.org/x/oauth2](https://github.com/golang/oauth2) from 0.21.0 to 0.27.0.
- [Commits](https://github.com/golang/oauth2/compare/v0.21.0...v0.27.0)

---
updated-dependencies:
- dependency-name: golang.org/x/oauth2
  dependency-version: 0.27.0
  dependency-type: indirect
...

Signed-off-by: dependabot[bot] <support@github.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update pre-commit hook renovatebot/pre-commit-hooks to v41.40.0

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): implement Kopia API types and mover types

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): implement tests for Kopia

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): start Kopia documentation

asfd

asfd

asdf

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): implement additional fields for kopia api types

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): update the kopia mover entry.sh for execute_action

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): format kopia files for linting, reduce complexity to pass lint

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): update description for kopia api objects

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add documentation examples for kopia usage

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): catch capital "FATAL" in log filter

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add kopia e2e tests, and actually have them pass

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): bump golang version and set correct kopia version for dockerfile

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): hopefully finish the kopia usage docs

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add kopia override support to the helm chart and run-in-kind.sh hack script

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add missing environment variables needed for kopia cmd

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): update entry script to get repository prefix, and add some light debug output

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): remove annoying log line about disabling tls

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add documentation around optional overrides for convenience

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): allow for username/hostname overrides in kopia repositories

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): create tests for optional overrides, and for multitenancy

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add Promtheus metrics for kopia objects

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add example PrometheusRule for Kopia

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): update documentation for kopia metrics

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): also allow for overriding hostnames

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(docs): implement temporary documentation action

asfd

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(docs): override the warning flag for releasing docs

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add environment variables for missing kopia backends

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): implement packaging the helm chart with the docs job

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): update readme to clarify that it's a fork and where to get the information that the user needs

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add documentation for all providers

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): mount credentials for gcs and sftp correctly

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): update readme to specify mover image too

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): tests now pass after fixing imports

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support manual config options

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): increment chart version

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(ci): disable additional unused jobs

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): resolve bug due to kopia prometheus metrics

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): allow the user to not have to define KOPIA_S3_BUCKET and just base it off of the S3 URL provided

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support even more environment variables

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add some more tests for kopia

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): don't print access keys to logs for debug anymore

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): check if variables are set or not set

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): use replicationsource name as username

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

renovate-bot committed:
Update dependency golangci/golangci-lint to v2.3.1

Signed-off-by: Mend Renovate <bot@renovateapp.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support not specifying a cache PVC, and implement fallback scenarios

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): update metrics and tests for new emptydir cache type

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(docs): add grafana dashboard example

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): update README for fork

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): update example kopia alert metrics

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): add missing `jq` binary to dockerfile

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support repositoryConfig object type

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): implement repositoryConfig for Kopia mover

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add repositoryConfig CRD to bundle manifests

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(controller): add Finalizers to stop sync loops

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): use ReplicationSource name for username, and namespace for hostname by default

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support previous field for ReplicationDestination

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): split up huge index docs page into multiple

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support sourceIdentity CRD for easy restoration

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): update readme examples as well

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): add another example

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): also update readme for container

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): also support sourceIdentity tests

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): resolve duplicated function in entry mover

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): resolve broken jq parsing for snapshots

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): add sourceIdentity for Kopia

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): update helm chart version

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support sourcePVC and the discovery of it for ReplicationDestinations

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add sourcePVC docs to sourceIdentity docs

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): add to `kopia snapshot restore` command to avoid unneeded atomic files

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): use the replicationdestination namespace if one isn't defined to get sourceIdentifier

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): update the entry script for listing snapshots

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): increment the helm version again

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add working autodiscovery tests

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): trailing slash fun

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): try to squash the weird `//data.kopia-entry` error

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): since it keeps trying to write in the parent directory, let's mount an emptydir above the data restore path

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): update the data path that it looks for in snapshots

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): update readme for working example

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support sourcePathOverride in ReplicationDestination

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): bump helm chart version for updated CRDs too

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): fix the mount path depending on if it's a source or destination

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): also "smartly" discover the `kopia.repository` from `sourceIdentity` if it isn't provided in `ReplicationDestination`

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): increment the helm chart

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(docs): update docs for sourceIdentity

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): update readme example

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): also support `filesystemDestination` for ReplicationSource

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(ci): recreate old Helm Chart versions because I'm dumb

asfd

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): use `repositoryPVC` instead of `filesystemDestination`

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): increment chart version

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): throw errors if requirements aren't met for ReplicationDestinations

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): also increment helm chart version

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(docs): clarify identity requirements for ReplicationDestinations

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): provide better metrics for passing/failing ReplicationSources

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(readme): add discord to readme

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): just have the `namespace` be the `hostname` in kopia backups

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): hostnames should just be the namespace

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): set up `enableFileDeletion` and docs for not having to specify `sourceIdentity`

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): have the username just be the `ReplicationSource` name

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): increment chart version

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): resolve issue with nil pointer dereference

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

onedr0p committed:
do not check for kopia updates from github

Signed-off-by: Devin Buhl <onedr0p@users.noreply.github.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

tesshuflower commented:
@perfectra1n hey thanks for this PR! Obviously a lot of work went into this - please be patient as it may take some time before we can get around to reviewing....

One thing just from glancing over the description:  The fix for "controller sync fix" - I think I'd prefer if there was an issue created for this and it was handled separately - I'm not sure I understand the need for a finalizer?  However it does sound like we most likely do need a check for the deletion timestamp.... An issue would help to put more info around what is getting stuck (for example is there a specific resource that is always getting stuck?) and putting in a targeted fix.

onedr0p committed:
if snapshots don't exist don't error during restore

Signed-off-by: Devin Buhl <onedr0p@users.noreply.github.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(ci): add container building step

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(ci): use a matrix instead for the container

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(ci): use a matrix instead for the container

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

onedr0p committed:
maybe this is the right way to check if no snapshots exist

Signed-off-by: Devin Buhl <onedr0p@users.noreply.github.com>
Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): remove finalizers, and per onedr0p: "delaying my testing ability at the moment because I need to manually rm resources"

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): increment helm version

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): further support global kopia repository configs

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): update docs for modifying kopia repository configs

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support the specification of compression in kopia repositories and/or snapshots

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): increment helm chart version

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(docker): try to use bfd instead of gold in arm64

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): fix incorrect compression statement

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): no longer validate compression input

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): update comments on volsync crds for kopia

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): support the use of additional arguments for kopia

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): add docs for additional args

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(helm): increment version for additional args

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): allow all args

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

tesshuflower commented:
> Sure! I can remove that from this PR. It was happening to my resources during development and in my homelab - see below: ![chrome_GROY6oQzw2](https://private-user-images.githubusercontent.com/2390633/477164509-60289314-4eb6-44a8-8ac1-d1315c4d0a18.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU3MTY2ODgsIm5iZiI6MTc1NTcxNjM4OCwicGF0aCI6Ii8yMzkwNjMzLzQ3NzE2NDUwOS02MDI4OTMxNC00ZWI2LTQ0YTgtOGFjMS1kMTMxNWM0ZDBhMTguZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MjBUMTg1OTQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmM2NjhhNTJkYzQyMjcyZDM3MzM5OTQ3NWE3MTgzZDkwYTViNTRlOGE1MDU1Y2U2ZTJhYTk4YzliOGViMThmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.f1IkMmxZu8jj1SDcU3UCG3uCT_EWGhFnDsB3zvX51r8) [ ![chrome_GROY6oQzw2](https://private-user-images.githubusercontent.com/2390633/477164509-60289314-4eb6-44a8-8ac1-d1315c4d0a18.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU3MTY2ODgsIm5iZiI6MTc1NTcxNjM4OCwicGF0aCI6Ii8yMzkwNjMzLzQ3NzE2NDUwOS02MDI4OTMxNC00ZWI2LTQ0YTgtOGFjMS1kMTMxNWM0ZDBhMTguZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MjBUMTg1OTQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmM2NjhhNTJkYzQyMjcyZDM3MzM5OTQ3NWE3MTgzZDkwYTViNTRlOGE1MDU1Y2U2ZTJhYTk4YzliOGViMThmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.f1IkMmxZu8jj1SDcU3UCG3uCT_EWGhFnDsB3zvX51r8) ](https://private-user-images.githubusercontent.com/2390633/477164509-60289314-4eb6-44a8-8ac1-d1315c4d0a18.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU3MTY2ODgsIm5iZiI6MTc1NTcxNjM4OCwicGF0aCI6Ii8yMzkwNjMzLzQ3NzE2NDUwOS02MDI4OTMxNC00ZWI2LTQ0YTgtOGFjMS1kMTMxNWM0ZDBhMTguZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MjBUMTg1OTQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmM2NjhhNTJkYzQyMjcyZDM3MzM5OTQ3NWE3MTgzZDkwYTViNTRlOGE1MDU1Y2U2ZTJhYTk4YzliOGViMThmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.f1IkMmxZu8jj1SDcU3UCG3uCT_EWGhFnDsB3zvX51r8) [ ](https://private-user-images.githubusercontent.com/2390633/477164509-60289314-4eb6-44a8-8ac1-d1315c4d0a18.gif?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTU3MTY2ODgsIm5iZiI6MTc1NTcxNjM4OCwicGF0aCI6Ii8yMzkwNjMzLzQ3NzE2NDUwOS02MDI4OTMxNC00ZWI2LTQ0YTgtOGFjMS1kMTMxNWM0ZDBhMTguZ2lmP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDgyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTA4MjBUMTg1OTQ4WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NmM2NjhhNTJkYzQyMjcyZDM3MzM5OTQ3NWE3MTgzZDkwYTViNTRlOGE1MDU1Y2U2ZTJhYTk4YzliOGViMThmZCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.f1IkMmxZu8jj1SDcU3UCG3uCT_EWGhFnDsB3zvX51r8)
> 
> I added a Finalizer as that's the pattern that other Kubernetes resources follow for lifecycle management, and it allows the `ReplicationDestination` or `ReplicationSource` to enter a state where it doesn't continuously reconcile resources.
> 
> I just couldn't take it anymore since it was extremely annoying to deal with during development (and I've run into it in the past when using Volsync throughout the past few years), so I just lumped in a fix 😂
> 
> I'll remove it from this PR though and just keep the fix in my fork for now, so that people that are testing out this Kopia implementation don't have to deal with infinite sync loops if they delete a ReplicationSource/ReplicationDestination before it completes lol.

@perfectra1n can you create a separate issue with details?  I think this is something we'd like to fix, but would like more information on exactly what resources are holding things up, etc.  Are there pvcs getting stuck in terminating for a long time or something of the sort?

tesshuflower commented on Dockerfile:
Just to warn you, we are not yet ready to move to golang 1.24 - there's actually an issue with building on arm64 right now with golang 1.24.  I would also like to move to golang 1.24 separately from the mover PR.  So this would either have to wait, or you can start with a build that works with 1.23.

perfectra1n replied:
I did fix that in my GitHub Actions though, so I'm able to build on ARM64 via [this](https://github.com/perfectra1n/volsync/commit/e03cf485004a1868eecd5168a87bac1aedb61e56) change, but I'm not sure if that aligns.

tesshuflower replied:
ah, thanks for that - I may try using that myself as the golang issue  seems to still be unresolved for the moment.  At least that would unblock us for moving to 1.24.

tesshuflower commented on config/rbac/role.yaml:
Is this intentional?  These permissions are required and should not be removed from the role

tesshuflower commented:
Note that the DCO check is failing because the commits are not signed off - please see the instructions here on how to resolve: https://github.com/backube/volsync/pull/1723/checks?check_run_id=48522602445

perfectra1n committed:
fix(kopia): reverse the kopia snapshot selection so it shows the "most recent" first

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): also implement useful unit test to make sure kopia snapshot selection is correct

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): resolve conflicts

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): continue to modify files that aren't needed for upstream

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): continue to modify files that aren't needed for upstream

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): continue to modify files that aren't needed for upstream

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): continue to modify files that aren't needed for upstream

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n commented:
Alright, I'm done messing around with the rebase - I believe I've addressed the concerns you highlighted @tesshuflower. I can also remove the Dockerfile change for Golang 1.24 if you'd prefer to just keep it at Golang 1.23.

perfectra1n committed:
fix(kopia): add log level and retention controls to prevent cache PVC overflow

The Kopia mover logs were filling up the 1GB cache PVC due to high default
retention settings and verbose debug logging. This commit adds environment
variable controls for log configuration with sensible defaults:

- Set default file log level to 'warn' instead of 'debug'
- Limit log retention to 10 files and 24 hours by default
- Add environment variables for users to override these settings:
  - KOPIA_FILE_LOG_LEVEL (default: warn)
  - KOPIA_LOG_DIR_MAX_FILES (default: 10)
  - KOPIA_LOG_DIR_MAX_AGE (default: 24h)
  - KOPIA_CONTENT_LOG_DIR_MAX_FILES (default: 10)
  - KOPIA_CONTENT_LOG_DIR_MAX_AGE (default: 24h)

These settings prevent the cache PVC from filling up while still maintaining
useful logs for debugging when needed. Users can adjust these values through
their Kopia repository secret if different retention is required.

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): also change the default logging to be info, and create docs for logging settings

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(docs): update defaults in docs for kopia changes

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
fix(kopia): instead of setting env var, provide custom ca to s3 command

Signed-off-by: perf3ct <jonfuller2012@gmail.com>

perfectra1n committed:
feat(kopia): implement latest kopia changes in PR, including moverVolumes

perfectra1n committed:
feat(dev): merge upstream main into pr

perfectra1n commented:
PR birthday party at RedHat HQ in August?

perfectra1n commented:
Hey all, I'm building [Kopiur](https://github.com/home-operations/kopiur) as a 1st class Kopia backup operator. There probably won't be any more changes to my fork (for now), especially since it's pretty thoroughly tested/proven now.

perfectra1n commented:
Hi all, I've moved on to creating and using [Kopiur](https://github.com/home-operations/kopiur) instead, so I'll be closing this PR. Please feel free to re-open and modify as needed Volsync contributors, if this is something you'd like to support in the future.
